Проект по SQL
=

Содержание
==

[Описание проекта](#0)

[Описание данных](#1)

[Схема данных](#2)

[Задачи](#3)

[1 Импортируем библиотеки](#4) 

[2 Создаём коннекцию к базам и получаем данные](#5) 

[3 Изучение данных](#6)

[4 Подготовка данных](#7)

[5 Изучение данных](#8)
- [5.1 Посчитаем, сколько книг вышло после 1 января 2000 года](#8.1)
- [5.2 Для каждой книги посчитаем количество обзоров и среднюю оценку](#8.2)
- [5.3 Определим издательства, которое выпустило наибольшее число книг](#8.3)
- [5.4 Определим автора с самой высокой средней оценкой книги](#8.4)
- [5.5 Посчитаем выручку на покупателя](#8.5)


[6 Вывод](#9)
- [6.1 Рекомендачии](#9.1)


Описание проекта
==
<a name="0"></a>

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.


**Описание данных**
===

<a name="1"></a>

**Таблица `books`**

Содержит данные о книгах:

* `book_id` — идентификатор книги;
* `author_id` — идентификатор автора;
* `title` — название книги;
* `num_pages` — количество страниц;
* `publication_date` — дата публикации книги;
* `publisher_id` — идентификатор издателя.


**Таблица `authors`**
Содержит данные об авторах:
* `author_id` — идентификатор автора;
* `author` — имя автора.

**Таблица `publishers`**
Содержит данные об издательствах:
* `publisher_id` — идентификатор издательства;
* `publisher` — название издательства;

**Таблица `ratings`**
Содержит данные о пользовательских оценках книг:
* `rating_id` — идентификатор оценки;
* `book_id` — идентификатор книги;
* `username` — имя пользователя, оставившего оценку;
* `rating` — оценка книги.

**Таблица ``reviews``**
Содержит данные о пользовательских обзорах:
* `review_id` — идентификатор обзора;
* `book_id` — идентификатор книги;
* `username` — имя автора обзора;
* `text` — текст обзора.



Схема данных
===

<a name="2"></a>

![схема данных](https://pictures.s3.yandex.net/resources/scheme_1589269096.png "Yandex")


Задачи
===
<a name="3"></a>


- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


## Импортируем библиотеки
<a name="4"></a>

In [1]:
import pandas as pd
from sqlalchemy import create_engine
from datetime import datetime
pd.set_option('display.max_columns', None)
pd.options.display.max_colwidth = 150

#import seaborn as sns; sns.set()
#cm = sns.light_palette("green", as_cmap=True)

#import seaborn as sns
#import math
#import numpy as np
#from scipy import stats as st

#import plotly.express as px
#import plotly.graph_objs as go
#from plotly.offline import download_plotlyjs, init_notebook_mode, plot, iplot

%autosave 60

Autosaving every 60 seconds


## Создаём коннекцию к базам и получаем данные
<a name="5"></a>

### books

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
# получаем сырые данные
query = '''
            SELECT * FROM books
        '''
books = pd.io.sql.read_sql(query, con = engine)

In [4]:
# формируем csv-файл
#books.to_csv('books.csv', sep='\t', encoding='utf-8', index=False)
books.head()

book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### authors

In [5]:
# получаем сырые данные
query = '''
            SELECT * FROM authors
        '''
authors = pd.io.sql.read_sql(query, con = engine)

In [6]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### publishers

In [7]:
# получаем сырые данные
query = '''
            SELECT * FROM publishers
        '''
publishers = pd.io.sql.read_sql(query, con = engine)

In [8]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### ratings

In [9]:
# получаем сырые данные
query = '''
            SELECT * FROM ratings
        '''
ratings = pd.io.sql.read_sql(query, con = engine)

In [10]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### reviews

In [11]:
# получаем сырые данные
query = '''
            SELECT * FROM reviews
        '''
reviews = pd.io.sql.read_sql(query, con = engine)

In [12]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.

## Изучение данных
<a name="6"></a>

In [13]:
list_df = (books, authors, publishers, ratings, reviews)
for list_df in list_df:
    print('\033[31m' + '______________info_____________________' + '\033[0m')
    display(list_df.info())
    print('\033[31m' + '______________describe_________________' + '\033[0m')
    display(list_df.describe().T)
#    print('\033[31m' + '______________head_____________________' + '\033[0m')
#    display(list_df.head(10))
    print('\033[31m' + '______________sample___________________' + '\033[0m')
    display(list_df.sample(10))
#    print('\033[31m' + '______________tail_____________________' + '\033[0m')
#    display(list_df.tail(10))
    print('\033[31m' + '_____________isna______________________' + '\033[0m')
    display(list_df.isna().sum().reset_index())
    print('\033[31m' + '_______Поиск дубликатов________________' + '\033[0m')
    display(list_df.duplicated().sum())
    print('\033[31m' + '__количество пропущенных значений по каждому из столбцов__' + '\033[0m')
    display(list_df.isnull().mean().reset_index())
    print('\033[34m' + '_________________________конец изучения файла____________________________________' + '\033[0m')
    print('')

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

______________describe_________________


count     mean         std   min     25%    50%     75%     max
book_id       1000.0  500.500  288.819436   1.0  250.75  500.5  750.25  1000.0
author_id     1000.0  320.417  181.620172   1.0  162.75  316.5  481.00   636.0
num_pages     1000.0  389.111  229.390140  14.0  249.00  352.0  453.00  2690.0
publisher_id  1000.0  171.270   99.082685   1.0   83.00  177.5  258.00   340.0

______________sample___________________


book_id  author_id  \
243      244        295   
273      274        372   
215      216        567   
102      103         28   
899      900        437   
719      720        378   
577      578        571   
522      523         14   
133      134        355   
533      534        235   

                                                               title  \
243                                                Faust  First Part   
273                                   Gathering Blue (The Giver  #2)   
215                                     Emperor Mage (Immortals  #3)   
102                                                     Black Beauty   
899                                      The Wake (The Sandman  #10)   
719                                                The Far Pavilions   
577                                      Small Gods (Discworld  #13)   
522                             Practical Magic (Practical Magic #1)   
133                                                       Cane River   
533  Raise High the Roof Beam  Carpenters & Seymour: An Introduction   

     num_pages publication_date  publisher_id  
243        327       1988-07-01            38  
273        240       2000-09-25            79  
215        358       2005-06-01           274  
102        245       2003-03-01           256  
899        191       1999-09-03           305  
719        958       1997-01-15           282  
577        400       2005-08-01            65  
522        286       2003-08-05           210  
133        522       2002-04-01           116  
533        256       2001-01-30            31

_____________isna______________________


index  0
0           book_id  0
1         author_id  0
2             title  0
3         num_pages  0
4  publication_date  0
5      publisher_id  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0           book_id  0.0
1         author_id  0.0
2             title  0.0
3         num_pages  0.0
4  publication_date  0.0
5      publisher_id  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

______________describe_________________


count   mean         std  min     25%    50%     75%    max
author_id  636.0  318.5  183.741666  1.0  159.75  318.5  477.25  636.0

______________sample___________________


author_id  \
595        596   
400        401   
9           10   
375        376   
43          44   
266        267   
359        360   
469        470   
222        223   
49          50   

                                                                     author  
595                                Upton Sinclair/Earl Lee/Kathleen DeGrave  
400                                                     Mark Z. Danielewski  
9                                      Aleksandr Solzhenitsyn/H.T. Willetts  
375                                                       Louisa May Alcott  
43                                                                 Ayn Rand  
266                                                           Jane Hamilton  
359                                                       Lauren Weisberger  
469                              Paulo Coelho/Amanda Hopkinson/Nick Caistor  
222                                           Hiromu Arakawa/Akira Watanabe  
49   Barbara Kingsolver/Steven L. Hopp/Camille Kingsolver/Richard A. Houser

_____________isna______________________


index  0
0  author_id  0
1     author  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  author_id  0.0
1     author  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

______________describe_________________


count   mean        std  min    25%    50%     75%    max
publisher_id  340.0  170.5  98.293777  1.0  85.75  170.5  255.25  340.0

______________sample___________________


publisher_id                       publisher
194           195                          Norton
12             13                   Allen & Ulwin
233           234                          Puffin
144           145  Harpercollins Children's Books
207           208                         Pearson
294           295           Top Shelf Productions
56             57      Cambridge University Press
209           210                         Penguin
99            100           Ecco/Harper Perennial
30             31                  Back Bay Books

_____________isna______________________


index  0
0  publisher_id  0
1     publisher  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  publisher_id  0.0
1     publisher  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

______________describe_________________


count         mean          std  min      25%     50%      75%  \
rating_id  6456.0  3228.500000  1863.831001  1.0  1614.75  3228.5  4842.25   
book_id    6456.0   510.574195   284.141636  1.0   291.00   506.0   750.00   
rating     6456.0     3.928284     0.943303  1.0     3.00     4.0     5.00   

              max  
rating_id  6456.0  
book_id    1000.0  
rating        5.0

______________sample___________________


rating_id  book_id       username  rating
3176       3177      497    sfitzgerald       5
1421       1422      251        qmiller       5
4493       4494      709        pthomas       4
3993       3994      655        debra43       2
853         854      144  timothygibson       4
5259       5260      813     ryanfranco       4
247         248       52        vcarter       2
1555       1556      284        fnguyen       5
5260       5261      813        carla64       3
2642       2643      414        odavila       5

_____________isna______________________


index  0
0  rating_id  0
1    book_id  0
2   username  0
3     rating  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  rating_id  0.0
1    book_id  0.0
2   username  0.0
3     rating  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

______________describe_________________


count         mean         std  min    25%     50%     75%     max
review_id  2793.0  1397.000000  806.413976  1.0  699.0  1397.0  2095.0  2793.0
book_id    2793.0   504.693161  288.472931  1.0  259.0   505.0   753.0  1000.0

______________sample___________________


review_id  book_id        username  \
292         291      100         nthomas   
1664       1663      603          xdavis   
291         290       99   michellewhite   
2240       2240      806     amymckenzie   
109         109       39     carlsonbrad   
495         495      181  nolanalexander   
1535       1534      557         wendy18   
1816       1815      657         carla64   
161         161       55   perrypatricia   
2151       2151      774       bobbyhood   

                                                                                                                                   text  
292          Store toward design. Each few wall investment claim top. End building staff itself hear put. It conference long about man.  
1664           Need discover old memory last plant. Bit first herself follow question throw. They right property president country our.  
291                                                   Sometimes watch lot dark cup always center. Cell bed city seat lead car discover.  
2240  Peace million project style. Entire customer message everything year increase population. Here less lead simple require consider.  
109                   Tough develop whole produce population still. Challenge kind lose large. Option set friend among figure computer.  
495                                                               President feel provide lay near may at. Produce into half result son.  
1535                                                  Black travel hospital season night ok. Speak oil education necessary range learn.  
1816                                                            Region billion do action trouble movie seek. Us buy sort money manager.  
161               Fly agree after behavior. Film focus ability include affect everyone. Environment approach bill baby nice face while.  
2151     Impact forget both responsibility old bit. Close either send analysis quality guy those teach. End wear child step be federal.

_____________isna______________________


index  0
0  review_id  0
1    book_id  0
2   username  0
3       text  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  review_id  0.0
1    book_id  0.0
2   username  0.0
3       text  0.0

_________________________конец изучения файла____________________________________



<a name="1.2"></a>
**Вывод:** 
* данные содержатся в 5 таблицах:
    - books 1000 строк и 6 колонок
    - authors 636 строк и 2 колонки
    - publishers 340 строк и 2 колонки
    - ratings 6456 строк и 4 колонки
    - reviews 2793 строки и 4 колонки
* явных дубликатов нет
* пропусков нет

## Подготовка данных
<a name="7"></a>

Изменим тип данных га соответствующий

In [14]:
books['publication_date'] = pd.to_datetime(books['publication_date'])

In [15]:
books.dtypes

book_id                      int64
author_id                    int64
title                       object
num_pages                    int64
publication_date    datetime64[ns]
publisher_id                 int64
dtype: object

Найдем скрытые дубликаты по автору и названию книги: 

In [16]:
books[['author_id', 'title']].duplicated().sum()

1

- Видим 1 дубликат, возможно это переиздание книги. Не будем обращать на это внимание.

Создадим функцию для запросов к данным:

In [17]:
def select(sql):
    return pd.read_sql(sql, con=engine)

## Изучение данных
<a name="8"></a>

### Посчитаем, сколько книг вышло после 1 января 2000 года
<a name="8.1"></a>

In [18]:
books_count = '''
                SELECT 
                COUNT(title) AS books_count
                FROM
                    books
                WHERE publication_date >= '2000-01-01';
             ''' 
select(books_count)

books_count
0          821

- Видим, что после 1 января 2000 г. вышла 821 книга

### Для каждой книги посчитаем количество обзоров и среднюю оценку
<a name="8.2"></a>

In [19]:
query_cnt_avg = '''                              
                    SELECT
                        books.title AS book_title,
                        books.book_id AS book_id,
                        AVG(ratings.rating) AS avg_rating,
                        COUNT(DISTINCT reviews.review_id) AS count_reviews -- DISTINCT только уники по столбцу

                    FROM
                        books

                    INNER JOIN ratings ON ratings.book_id = books.book_id
                    LEFT JOIN reviews ON reviews.book_id = books.book_id
                    

                    GROUP BY
                        books.title,
                        books.book_id

                    ORDER BY
                        count_reviews DESC;
                ''' 
select(query_cnt_avg).round(3)#.head(10)

book_title  book_id  \
0                                        Twilight (Twilight  #1)      948   
1                                            Water for Elephants      963   
2                                               The Glass Castle      734   
3    Harry Potter and the Prisoner of Azkaban (Harry Potter  #3)      302   
4              The Curious Incident of the Dog in the Night-Time      695   
..                                                           ...      ...   
995              Anne Rice's The Vampire Lestat: A Graphic Novel       83   
996                                      The Natural Way to Draw      808   
997             The Cat in the Hat and Other Dr. Seuss Favorites      672   
998                                    Essential Tales and Poems      221   
999         Disney's Beauty and the Beast (A Little Golden Book)      191   

     avg_rating  count_reviews  
0         3.662              7  
1         3.977              6  
2         4.207              6  
3         4.415              6  
4         4.081              6  
..          ...            ...  
995       3.667              0  
996       3.000              0  
997       5.000              0  
998       4.000              0  
999       4.000              0  

[1000 rows x 4 columns]

Проверим количество рецензий по столбцу `count_reviews`

In [20]:
check = select(query_cnt_avg)
check_rew = check.count_reviews.sum()
check_rew

2793

- Определили для каждой книги количество обзоров и среднюю оценку

### Определим издательства, которое выпустило наибольшее число книг
<a name="8.3"></a>

- По условию книги все что больше 50 страниц

In [21]:
query_publisher = '''
                     SELECT 
                         publishers.publisher_id, 
                         publishers.publisher, 
                         COUNT(books.book_id) as count_books
                     FROM 
                         publishers
                     JOIN books ON publishers.publisher_id = books.publisher_id
                     WHERE 
                         books.num_pages > 50
                     GROUP BY publishers.publisher_id
                     ORDER BY count_books DESC
                     LIMIT 10;
                 ''' 
select(query_publisher)

publisher_id                 publisher  count_books
0           212             Penguin Books           42
1           309                   Vintage           31
2           116  Grand Central Publishing           25
3           217          Penguin Classics           24
4            33          Ballantine Books           19
5            35                    Bantam           19
6            45                   Berkley           17
7           284        St. Martin's Press           14
8            46             Berkley Books           14
9            83                     Delta           13

- Больше всего книг выпустило издательство `Penguin Books`

### Определим автора с самой высокой средней оценкой книги
<a name="8.4"></a>

- По условию учитываем книги с 50 и более оценками

In [22]:
query_rating_avg = '''
                        SELECT
                           authors.author_id AS id,
                           authors.author AS author,
                           AVG(ratings.rating) AS avg_rating,
                           COUNT(ratings.rating) AS count_rating
                        FROM
                           authors
                        INNER JOIN books ON books.author_id = authors.author_id
                        INNER JOIN ratings ON ratings.book_id = books.book_id
                        GROUP BY
                           id,
                           author
                        HAVING 
                           COUNT(ratings.rating) >= 50 
                           
                        ORDER BY
                           count_rating DESC
                        LIMIT 1
                        ;
                '''
select(query_rating_avg).round(3)

id                      author  avg_rating  count_rating
0  236  J.K. Rowling/Mary GrandPré       4.288           312

- По результатам видим, что автор с самой высокой средней оценкой книг (с 50 и более оценками): `J.K. Rowling`

### Посчитаем среднее количество обзоров от пользователей
<a name="8.5"></a>

- По условию учитываем пользователей, которые поставили больше 50 оценок.

In [23]:
query_publisher_avg = '''
                        SELECT
                        AVG(review.review_cnt) AS avg_review
                        FROM
                            (SELECT COUNT(review_id) AS review_cnt
                        FROM 
                            reviews
                        WHERE
                            username
                        IN 
                            (SELECT username
                        FROM
                            ratings
                        GROUP BY
                            username
                        HAVING
                        COUNT(ratings) > 50)
                        GROUP BY 
                            username) 
                        AS
                        review;
                     '''
select(query_publisher_avg).round(0)

avg_review
0        24.0

- Пользователи оставившие более 50 оценок, в среднем написали 24 обзора

## Вывод
<a name="9"></a>

* В ходе исследования  проанализировать базу данных сервиса чтения книг.

* Исследования проводились на основе БД, включающей из 5 таблиц:
    - `books`
    - `authors`
    - `publishers`
    - `ratings`
    - `reviews`
    
* В процессе исследования определили что
    - после 1 января 2000 г. вышла 821 книга
    - для каждой книги о пределиликоличество обзоров и среднюю оценку
    - больше всего книг выпустило издательство `Penguin Books`
    - автор с самой высокой средней оценкой книг (с 50 и более оценками): `J.K. Rowling`
    - пользователи оставившие более 50 оценок, в среднем написали 24 обзора



### Рекомендации
<a name="9.1"></a>

* Для максимизации прибыли следует ориентироваться на бестселлеры и (или) расширение предложения за счет специальзированной литературы (возможно авторских тематических публикаций)

* Также стоит рассмотреть вариант расширения функционала (и предложения) за счет аудиокниг (тематических подкастов), которые пользуются популярностью у определеннной группы людей